app/src/ui/rebase/rebase-branch-dialog.tsx

```diff
@@ -0,0 +1,189 @@
+
+import { Dialog, DialogContent, DialogFooter } from '../dialog'
+import { BranchList, IBranchListItem, renderDefaultBranch } from '../branches'
+import { IMatches } from '../../lib/fuzzy-find'
+import { truncateWithEllipsis } from '../../lib/truncate-with-ellipsis'
+import { DialogHeader } from '../dialog/header'
+import { Loading } from '../lib/loading'
+
+interface IRebaseBranchDialogProps {
+  readonly dispatcher: Dispatcher
+  readonly repository: Repository
+
+  /**
+   * See IBranchesState.defaultBranch
+   */
+  readonly defaultBranch: Branch | null
+
+  /**
+   * The currently checked out branch
+   */
+  readonly currentBranch: Branch
+
+  /**
+   * See IBranchesState.allBranches
+   */
+  readonly allBranches: ReadonlyArray<Branch>
+
+  /**
+   * See IBranchesState.recentBranches
+   */
+  readonly recentBranches: ReadonlyArray<Branch>
+
+  /**
+   * The branch to select when the rebase dialog is opened
+   */
+  readonly initialBranch?: Branch
+
+  /**
+   * A function that's called when the dialog is dismissed by the user in the
+   * ways described in the Dialog component's dismissable prop.
+   */
+  readonly onDismissed: () => void
+}
+
+interface IRebaseBranchDialogState {
+  /** The currently selected branch. */
+  readonly selectedBranch: Branch | null
+
+  /** The filter text to use in the branch selector */
+  readonly filterText: string
+
+  readonly isRebasing: boolean
+}
+
+/** A component for initating a rebase of the current branch. */
+export class RebaseBranchDialog extends React.Component<
+  IRebaseBranchDialogProps,
+  IRebaseBranchDialogState
+> {
+  public constructor(props: IRebaseBranchDialogProps) {
+    super(props)
+
+    const selectedBranch = this.resolveSelectedBranch()
+
+    this.state = {
+      selectedBranch,
+      filterText: '',
+      isRebasing: false,
+    }
+  }
+
+  private onFilterTextChanged = (filterText: string) => {
+    this.setState({ filterText })
+  }
+
+  private onSelectionChanged = (selectedBranch: Branch | null) => {
+    this.setState({ selectedBranch })
+  }
+
+  private renderBranch = (item: IBranchListItem, matches: IMatches) => {
+    return renderDefaultBranch(item, matches, this.props.currentBranch)
+  }
+
+  public render() {
+    const selectedBranch = this.state.selectedBranch
+    const currentBranch = this.props.currentBranch
+
+    const selectedBranchIsNotCurrentBranch =
+      selectedBranch === null ||
+      currentBranch === null ||
+      currentBranch.name === selectedBranch.name
+
+    const disabled = selectedBranchIsNotCurrentBranch
+
+    // the amount of characters to allow before we truncate was chosen arbitrarily
+    const currentBranchName = truncateWithEllipsis(
+      this.props.currentBranch.name,
+      40
+    )
+
+    const loading = this.state.isRebasing ? <Loading /> : null
+
+    return (
+      <Dialog
+        id="rebase"
+        onDismissed={this.props.onDismissed}
+        onSubmit={this.startRebase}
+      >
+        <DialogHeader
```

Comment:
Manually rendering a header like this will break the accessibility support built into the dialog probably causing screen readers to no longer announce the title of the dialog accurately. The `<dialog>` element is marked with an `aria-labelledby` attribute which points to the dialog header id (which is creating on the fly by the dialog component).

The way I'd prefer this to be implemented is by changing the signature of the `Dialog` component `title` prop to allow `string | JSX.Element` and passing the custom markup through that.

```tsx
<Dialog
  id="rebase"
  onDismissed={this.props.onDismissed}
  onSubmit={this.startRebase}
  title={
    <>
      Rebase <strong>{currentBranchName}</strong> onto…
    </>
  }
>
```

Reply:
This feels like an earlier discussion around this area #6032 - it's been a while since I looked at this prop but I'll give it another shot and see what comes out of it.

Reply:
Updated the API in 6cf4a2aab1a4a648c7becdd1c4b6777bd1534bb2 and migrated over to  7f4cedd5cd7d71b8cef35ae06a7c956a9715ded9